doc/_extensions/zephyr/domain.py

```diff
@@ -0,0 +1,287 @@
+                                                        reftarget=code_sample['id'], refwarn=True)
+                    sample_xref += nodes.inline(text=code_sample['name'])
+                    sample_para += sample_xref
+                    sample_para += nodes.inline(text=' - ')
+                    sample_para += nodes.inline(text=code_sample['description'].astext())
+                    sample_li = nodes.list_item()
+                    sample_li += sample_para
+                    sample_ul += sample_li
+                admonition += sample_ul
+
+                # replace node with the newly created admonition
+                node.replace_self(admonition)
+            else:
+                # remove node if there are no code samples
+                node.replace_self([])
+
+class CodeSampleDirective(Directive):
+    """
+    A directive for creating a code sample node in the Zephyr documentation.
+    """
+
+    required_arguments = 1  # ID
+    optional_arguments = 0
+    option_spec = {'name': directives.unchanged,
+                   'relevant-api': directives.unchanged}
+    has_content = True
+
+    def run(self):
+        code_sample_id = self.arguments[0]
+        env = self.state.document.settings.env
+        code_samples = env.domaindata["zephyr"]["code-samples"]
+
+        if code_sample_id in code_samples:
+            logger.warning(f"Code sample {code_sample_id} already exists. "
+                         f"Other instance in {code_samples[code_sample_id]['docname']}",
+                         location=(env.docname, self.lineno))
+
+        name = self.options.get('name', code_sample_id)
+        relevant_api_list = self.options.get('relevant-api', '').split()
+
+        # Create a node for description and populate it with parsed content
+        description_node = nodes.container(ids=[f"{code_sample_id}-description"])
+        self.state.nested_parse(self.content, self.content_offset, description_node)
+
+        code_sample = {
+            'id': code_sample_id,
+            'name': name,
+            'description': description_node,
+            'relevant-api': relevant_api_list,
+            'docname': env.docname
+        }
+
+        domain =env.get_domain("zephyr")
+        domain.add_code_sample(code_sample)
+
+        # Create an instance of the custom node
+        code_sample_node = CodeSampleNode()
+        code_sample_node['id'] = code_sample_id
+        code_sample_node['name'] = name
+
+        return [code_sample_node]
+
+class ZephyrDomain(Domain):
+    """Zephyr domain"""
+
+    name = 'zephyr'
+    label = 'Zephyr Project'
+
+    roles = {
+        'code-sample': XRefRole(innernodeclass=nodes.inline),
+    }
+
+    directives = {
+        'code-sample': CodeSampleDirective
+    }
+
+    object_types: Dict[str, ObjType] = {
+        'code-sample': ObjType('code sample', 'code-sample'),
+    }
+
+    initial_data: Dict[str, Any] = {
+        "code-samples": {}
+    }
```

Comment:
nit: I'd suggest to run the `black` formatter, file has some inconsistencies, here for example mixed usage of `"` and `'`.

Reply:
oh, good call! The only formatting I applied was sorting the imports so yeah, will do!

Reply:
ran `black` -- looking much better know, thanks!